### PR TITLE
prov/sockets: Added reference counting for pollset

### PR DESCRIPTION
**Pull request by shantonu**
- Added reference counting for pollset
- Cleanup code

@jithinjosepkl
Signed-off-by: Shantonu Hossain <shantonu.hossain@intel.com>

### CODE REVIEW

File: prov/sockets/src/sock_poll.c [hunk 1]
```diff
@@ -55,6 +58,24 @@ static int sock_poll_add(struct fid_poll *pollset, struct fid *event_fid,
...
+	case FI_CLASS_CQ:
+		cq = container_of(list_item->fid, struct sock_cq, cq_fid);
+		atomic_inc(&cq->ref);
+		break;
+	case FI_CLASS_CNTR:
+		cntr = container_of(list_item->fid, struct sock_cntr, cntr_fid);
+		atomic_inc(&cntr->ref);
+		break;
+	case FI_CLASS_EQ:
+		eq = container_of(list_item->fid, struct sock_eq, eq);
```
Comment (jithinjosepkl) [inline prov/sockets/src/sock_poll.c]: We cannot add EQ to a pollset. EQ is a fabric object and pollset is a domain object.

File: prov/sockets/src/sock_eq.c [hunk 2]
```diff
@@ -268,6 +268,9 @@ static int sock_eq_fi_close(struct fid *fid)
 	struct sock_eq *sock_eq;
 
 	sock_eq = container_of(fid, struct sock_eq, eq);
+	if (atomic_get(&sock_eq->ref))
+                return -FI_EBUSY;
+
```
Comment (jithinjosepkl) [inline prov/sockets/src/sock_eq.c]: Can we add ref-counting for EQs also? (May be as a separate commit.)

File: prov/sockets/src/sock_poll.c (hunk 1 above)
Comment (shefty) [inline prov/sockets/src/sock_poll.c]: Does the code take reference counts in both directions?  I.e. the CQ on the poll set and the poll set on the CQ.

Comment (shantonu) [inline prov/sockets/src/sock_poll.c]: @jithinjosepkl, didn't notice that. will remove then.
@shefty, no, it doesn't. It does reference counting for domain.

File: prov/sockets/src/sock_poll.c [hunk 3]
```diff
@@ -151,6 +182,18 @@ static int sock_poll_close(fid_t fid)
...
+			cq = container_of(list_item->fid, struct sock_cq, cq_fid);
+			atomic_dec(&cq->ref);
+			break;
+		case FI_CLASS_CNTR:
+			cntr = container_of(list_item->fid, struct sock_cntr, cntr_fid);
+			atomic_dec(&cntr->ref);
+			break;
+		default:
+			break;
+		}
```
Comment (shefty) [inline prov/sockets/src/sock_poll.c]: This code segment is shared with sock_poll_del -- please merge to common code, or call sock_poll_del directly from sock_poll_close.

Comment (jithinjosepkl) [inline prov/sockets/src/sock_poll.c]: Also, it would be nice to handle the default case (error log/return -FI_EINVAL).

Comment (shantonu) [inline prov/sockets/src/sock_poll.c]: Done!

**Comment by shantonu** [comment]
Updated with the feedbacks.

**Comment by shefty** [comment]
:+1:

**Comment by jithinjosepkl** [comment]
:+1:
